extensions/markdown/src/extension.ts

```diff
@@ -281,13 +298,36 @@ class MDDocumentContentProvider implements vscode.TextDocumentContentProvider {
 			}, 300);
 		}
 	}
+}
+
+class MDDocumentSymbolProvider implements vscode.DocumentSymbolProvider {
+
+	constructor(private markdownIt: MarkdownIt) { }
 
-	private getDocumentContentForPreview(document: vscode.TextDocument): string {
-		const content = document.getText();
-		const previewFrontMatter = vscode.workspace.getConfiguration('markdown')['previewFrontMatter'];
-		if (previewFrontMatter === 'hide') {
-			return content.replace(/^-{3}[ \t]*(\r\n|\n)(.|\r\n|\n)*?(\r\n|\n)-{3}[ \t]*(\r\n|\n)/, '');
+	provideDocumentSymbols(document: vscode.TextDocument): vscode.ProviderResult<vscode.SymbolInformation[]> {
+		let offset = 0;
+		let text = document.getText();
+		const frontMatterMatch = FrontMatterRegex.exec(text);
+
+		if (frontMatterMatch) {
+			const frontMatter = frontMatterMatch[0];
+
+			offset = frontMatter.split(/\r\n|\n|\r/g).length - 1;
+			text = text.substr(frontMatter.length);
 		}
-		return content;
+
+		const tokens = this.markdownIt.parse(text);
+		const headings = tokens.filter(token => token.type === 'heading_open');
+
+		const symbols = headings.map(heading => {
+			const lineNumber = heading.map[0];
+			const line = document.lineAt(lineNumber + offset);
+			const location = new vscode.Location(document.uri, line.range);
+			const text = line.text.replace(/^\s*#+\s*/, '');
```

Comment:
In order to handle headings such as `# a #`, perform the same replacement backwards on the end of the string too (the number of leading `#` does not have to match the number of trailing `#`)